File: week-03_05/tp/tp/tp6/commsplit.f90
program commsplit
  use MPI_F08
  implicit none
  TYPE(MPI_Comm)                 :: CommCart2D,CommCart1D
  integer, parameter             :: NDimCart2D=2
  integer, dimension(NDimCart2D) :: DimCart2D,CoordCart2D
  logical, dimension(NDimCart2D) :: Periods
  logical                        :: Reorder
  integer                        :: nb_procs,rank,i
  integer, parameter             :: m=4
  real, dimension(m)             :: V
  real                           :: W

  call MPI_INIT()
  call MPI_COMM_RANK(MPI_COMM_WORLD, rank)
  call MPI_COMM_SIZE( MPI_COMM_WORLD, nb_procs)

  !*** 2D Cartesian topology
  DimCart2D(1) = 4
  DimCart2D(2) = 2
  if (DimCart2D(1)*DimCart2D(2) /= nb_procs) then
    print *, "The number of processes is not correct !"
    ! TODO: On arrete le programme

  end if

  Periods(:)   = .false.
  Reorder      = .false.

  ! TODO: Creation of the CommCart2D communicator (2D cartesian topology)

  ! TODO: Get coordinates in topology -> CoordCart2D

  !*** Initialization of the V array and the W scalar
  V(:) = 0.
  W = 0.
  if (CoordCart2D(1) == 1) V(:) = (/ (real(i), i=1,m) /)

  ! TODO: Subdividing the 2D cartesian topology with MPI_COMM_SPLIT
  

  ! TODO: The processes of the second column scatter the V array
  ! to processes of their line
  

  print '("Rank : ",I2," ; Coordinates : (",I1,",",I1,") ; W = ",F2.0)', &
        rank,CoordCart2D(1),CoordCart2D(2),W

  ! TODO: Destruction of the communicators
  
  call MPI_FINALIZE()
end program commsplit

File: week-03_05/tp/tp_C/tp6/commsplit.c
#include <mpi.h>
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char *argv[]) {
  int nb_procs,rank;
  const int nDimCart2D=2;
  int dimCart2D[nDimCart2D];
  int periods[nDimCart2D];
  const int no=0;
  int reorder;
  MPI_Comm commCart2D,commCart1D;
  int coordCart2D[nDimCart2D];
  const int m=4;
  double v[m];
  double w;

  MPI_Init( &argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);
  MPI_Comm_size(MPI_COMM_WORLD, &nb_procs);

  dimCart2D[0] = 4;
  dimCart2D[1] = 2;
  if (dimCart2D[0]*dimCart2D[1] != nb_procs) {
    fprintf(stderr, "The number of processes is not correct !");
    /* TODO: We stop the program */

  }
  periods[0] = no; periods[1] = no;
  reorder = no;

  /* TODO: Creation of commcart2D comunicator (2D cartesian topology) */

  /* TODO: Get the coordinates of the current process */

  /* Initialization of V array and W scalar */
  w = v[0] = v[1] = v[2] = v[3] = 0;
  if (coordCart2D[0] == 1) {
    v[0] = 1; v[1] = 2; v[2] = 3; v[3] = 4; }

  /* TODO: Subdividing the 2D cartesian topology with MPI_COMM_SPLIT */

  /* TODO: Process of the second column scatters the V array
   * to each process of their line */

  printf("rank : %2d ; Coordinates ( %1d , %1d ) ; W = %2f\n",
         rank, coordCart2D[0], coordCart2D[1], w);

  MPI_Comm_free(&commCart1D);
  MPI_Comm_free(&commCart2D);
  MPI_Finalize();
  return 0;
}

